File: examples/elixir/basic-server/test/basic_service_rest_json_1_test.exs
defmodule BasicServiceRestJson1Test do
  use ExUnit.Case, async: true

  alias BasicTypes.GetTypeClosureOutput
  alias RuntimeTypes.HttpRequest

  describe "decode_get_type_closure_request/2" do
    test "minimal request" do
      req = %HttpRequest{
        method: "GET",
        path: "/types/widget",
        query: %{},
        headers: [{"Content-Type", "application/json"}],
        body: ""
      }

      label_map = %{"name" => "widget"}
      input = BasicServiceRestJson1.decode_get_type_closure_request(req, label_map)
      assert input.name == "widget"
      assert input.verbose == nil
      assert input.request_tag == nil
    end

    test "full request with optional fields" do
      req = %HttpRequest{
        method: "GET",
        path: "/types/widget",
        query: %{"verbose" => "true"},
        headers: [{"X-Request-Tag", "trace-1"}],
        body: ""
      }

      label_map = %{"name" => "widget"}
      input = BasicServiceRestJson1.decode_get_type_closure_request(req, label_map)
      assert input.name == "widget"
      assert input.verbose == true
      assert input.request_tag == "trace-1"
    end

    test "URI-decodes path label" do
      req = %HttpRequest{
        method: "GET",
        path: "/types/hello%20world",
        query: %{},
        headers: [],
        body: ""
      }

      label_map = %{"name" => "hello world"}
      input = BasicServiceRestJson1.decode_get_type_closure_request(req, label_map)
      assert input.name == "hello world"
    end
  end

  describe "encode_get_type_closure_response/1" do
    test "success with empty body" do
      out = %GetTypeClosureOutput{etag: "\"v1\"", basic_string: nil}

      resp = BasicServiceRestJson1.encode_get_type_closure_response(out)
      assert resp.status == 200
      assert resp.body == "{}"
    end

    test "success with JSON body" do
      out = %GetTypeClosureOutput{
        etag: "\"etag\"",
        basic_string: "hello",
        basic_integer: 42,
        basic_boolean: true
      }

      resp = BasicServiceRestJson1.encode_get_type_closure_response(out)
      assert resp.status == 200

      assert Jason.decode!(resp.body) == %{
               "basicString" => "hello",
               "basicInteger" => 42,
               "basicBoolean" => true
             }
    end
  end
end

File: examples/erlang/basic-server/test/basic_service_rest_json_1_test.erl
-module(basic_service_rest_json_1_test).

-include_lib("eunit/include/eunit.hrl").
-include("basic_service_types.hrl").
-include("runtime_types.hrl").

%% decode_get_type_closure_request/2

decode_get_type_closure_request_uses_label_map_test() ->
    Req = #http_request{
        method = <<"GET">>,
        path = <<"/types/widget">>,
        query = #{<<"verbose">> => <<"true">>},
        headers = [{<<"X-Request-Tag">>, <<"trace-1">>}],
        body = <<>>
    },
    LabelMap = #{<<"name">> => <<"widget">>},
    Input = basic_service_rest_json_1:decode_get_type_closure_request(Req, LabelMap),
    ?assertEqual(<<"widget">>, Input#get_type_closure_input.name),
    ?assertEqual(true, Input#get_type_closure_input.verbose),
    ?assertEqual(<<"trace-1">>, Input#get_type_closure_input.request_tag).

decode_minimal_request_test() ->
    Req = #http_request{
        method = <<"GET">>,
        path = <<"/types/widget">>,
        query = #{},
        headers = [{<<"Content-Type">>, <<"application/json">>}],
        body = <<>>
    },
    LabelMap = #{<<"name">> => <<"widget">>},
    Input = basic_service_rest_json_1:decode_get_type_closure_request(Req, LabelMap),
    ?assertEqual(<<"widget">>, Input#get_type_closure_input.name),
    ?assertEqual(undefined, Input#get_type_closure_input.verbose),
    ?assertEqual(undefined, Input#get_type_closure_input.request_tag).

decode_full_request_test() ->
    Req = #http_request{
        method = <<"GET">>,
        path = <<"/types/widget">>,
        query = #{<<"verbose">> => <<"true">>},
        headers = [{<<"X-Request-Tag">>, <<"trace-1">>}],
        body = <<>>
    },
    LabelMap = #{<<"name">> => <<"widget">>},
    Input = basic_service_rest_json_1:decode_get_type_closure_request(Req, LabelMap),
    ?assertEqual(<<"widget">>, Input#get_type_closure_input.name),
    ?assertEqual(true, Input#get_type_closure_input.verbose),
    ?assertEqual(<<"trace-1">>, Input#get_type_closure_input.request_tag).

decode_uri_decodes_path_label_test() ->
    Req = #http_request{
        method = <<"GET">>,
        path = <<"/types/hello%20world">>,
        query = #{},
        headers = [],
        body = <<>>
    },
    LabelMap = #{<<"name">> => <<"hello world">>},
    Input = basic_service_rest_json_1:decode_get_type_closure_request(Req, LabelMap),
    ?assertEqual(<<"hello world">>, Input#get_type_closure_input.name).

%% encode_get_type_closure_response/1

encode_get_type_closure_response_empty_body_test() ->
    Out = #get_type_closure_output{
        etag = <<"\"v1\"">>,
        basic_string = undefined
    },
    Resp = basic_service_rest_json_1:encode_get_type_closure_response(Out),
    ?assertEqual(200, Resp#http_response.status),
    ?assertEqual(jsone:encode(#{}), Resp#http_response.body).

encode_get_type_closure_response_json_body_test() ->
    Out = #get_type_closure_output{
        etag = <<"\"etag\"">>,
        basic_string = <<"hello">>,
        basic_integer = 42,
        basic_boolean = true
    },
    Resp = basic_service_rest_json_1:encode_get_type_closure_response(Out),
    ?assertEqual(200, Resp#http_response.status),
    ?assertEqual(
        jsone:encode(#{
            <<"basicString">> => <<"hello">>,
            <<"basicInteger">> => 42,
            <<"basicBoolean">> => true
        }),
        Resp#http_response.body
    ).
